File: 0189-rotate-array/0189-rotate-array.cpp
class Solution {
public:
    // void rotate(vector<int>& nums, int k) {
    //     int n=nums.size();
    //     vector<int>temp(n);
    //     int j=n-k;
    //     for(int i=0;i<n;i++){
    //        temp[i]=nums[j%n];
    //        j++;
    //     }
    //     nums=temp;
    // }

    void rotate(vector<int>& nums, int k) {
        int n=nums.size();
        k=k%n;
        reverse(nums.begin(),nums.end());
        reverse(nums.begin(),nums.begin()+k);
        reverse(nums.begin()+k,nums.end());
    }
};

File: 0189-rotate-array/0189-rotate-array.java
class Solution {

    void reverseTill(int[] nums,int s,int e){
        while(s<e){
            int temp=nums[s];
            nums[s++]=nums[e];
            nums[e--]=temp;
        }
    }

    public void rotate(int[] nums, int k) {
        int n=nums.length;
        k=k%n;
        if(n==0) return;
        reverseTill(nums,0,n-k-1);
        reverseTill(nums,n-k,n-1);
        reverseTill(nums,0,n-1);
    }
}